### PR TITLE
curvefs/metaserver: fixed rocksdb storage memory leak caused by unrel…

### PR DESCRIPTION
**Pull request by Wine93**
…easing iterator and transaction.

### What problem does this PR solve?

Issue Number: #1381

Problem Summary: rocksdb storage memory leak

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

File: curvefs/src/metaserver/storage/rocksdb_storage.h [hunk 1]
```diff
@@ -355,6 +361,10 @@ class RocksDBStorageIterator : public Iterator {
                 storage_->db_->ReleaseSnapshot(readOptions_.snapshot);
             }
         }
+
+        if (iter_ != nullptr) {
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/storage/rocksdb_storage.h:365]: wrap `iter_` with a std::unique_ptr?

File: curvefs/src/metaserver/storage/rocksdb_storage.h [hunk 2]
```diff
@@ -86,12 +86,14 @@ class RocksDBOptions {
     std::vector<ColumnFamilyDescriptor> columnFamilies_;
 
     static const std::string kOrderedColumnFamilyName_;
+
+    std::shared_ptr<rocksdb::Comparator> comparator_;
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/storage/rocksdb_storage.h:90]: there is no need to wrap it with a std::shared_ptr, just storing the raw value is enough.

File: curvefs/src/metaserver/storage/rocksdb_storage.cpp [hunk 3]
```diff
@@ -42,39 +42,43 @@ const std::string RocksDBOptions::kOrderedColumnFamilyName_ =  // NOLINT
...
     tableOptions.pin_l0_filter_and_index_blocks_in_cache = true;
-    tableOptions.filter_policy.reset(NewBloomFilterPolicy(10, true));
-    dbOptions_.table_factory.reset(NewBlockBasedTableFactory(tableOptions));
+    tableOptions.filter_policy.reset(NewBloomFilterPolicy(10, false));
 
     // column failmy options
-    auto unorderedCFOptions = ColumnFamilyOptions();
-    auto orderedCFOptions = ColumnFamilyOptions();
+    comparator_ = std::make_shared<RocksDBStorageComparator>();
+    auto cfOptions = ColumnFamilyOptions();
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/storage/rocksdb_storage.cpp:63]: `ColumnFamilyOptions cfOptions;`

File: curvefs/src/metaserver/storage/rocksdb_storage.cpp [hunk 4]
```diff
@@ -404,24 +408,28 @@ std::shared_ptr<StorageTransaction> RocksDBStorage::BeginTransaction() {
...
 Status RocksDBStorage::Rollback()  {
-    if (!InTransaction_) {
+    if (!InTransaction_ || nullptr == txn_) {
         return Status::NotSupported();
     }
     pending4set_.clear();
     pending4del_.clear();
-    return ToStorageStatus(txn_->Rollback());
+    Status s = ToStorageStatus(txn_->Commit());
+    delete txn_;
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/storage/rocksdb_storage.cpp:431]: ditto, use std::unique_ptr if possiable

File: curvefs/src/metaserver/storage/rocksdb_storage.cpp [hunk 5]
```diff
@@ -42,39 +42,43 @@ const std::string RocksDBOptions::kOrderedColumnFamilyName_ =  // NOLINT
...
     // column failmy options
-    auto unorderedCFOptions = ColumnFamilyOptions();
-    auto orderedCFOptions = ColumnFamilyOptions();
+    comparator_ = std::make_shared<RocksDBStorageComparator>();
+    auto cfOptions = ColumnFamilyOptions();
+    cfOptions.comparator = comparator_.get();
+    cfOptions.enable_blob_files = true;
+    cfOptions.level_compaction_dynamic_level_bytes = true;
+    cfOptions.compaction_pri = ROCKSDB_NAMESPACE::kMinOverlappingRatio;
+    cfOptions.prefix_extractor.reset(NewCappedPrefixTransform(3));
```
Comment (SeanHai) [inline curvefs/src/metaserver/storage/rocksdb_storage.cpp:68]: Does some options here important and add some comments？

File: curvefs/src/metaserver/storage/rocksdb_storage.cpp (hunk 4 above)
Comment (SeanHai) [reply curvefs/src/metaserver/storage/rocksdb_storage.cpp:431]: agree

File: curvefs/src/metaserver/storage/rocksdb_storage.cpp (hunk 5 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/rocksdb_storage.cpp:68]: > Does some options here important and add some comments？

fixed.

File: curvefs/src/metaserver/storage/rocksdb_storage.cpp (hunk 3 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/rocksdb_storage.cpp:63]: > `ColumnFamilyOptions cfOptions;`

fixed.

File: curvefs/src/metaserver/storage/rocksdb_storage.cpp (hunk 4 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/rocksdb_storage.cpp:431]: > agree

fixed.

File: curvefs/src/metaserver/storage/rocksdb_storage.h (hunk 2 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/rocksdb_storage.h:90]: > 

fixed.

File: curvefs/src/metaserver/storage/rocksdb_storage.h (hunk 1 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/rocksdb_storage.h:365]: > wrap `iter_` with a std::unique_ptr?

fixed.

**Review by wu-hanqing — APPROVED** [review]

**Review by SeanHai — APPROVED** [review]

**Pull request merged**